Document adding concepts to an existing schema

izmalk opened the pull request:
## What is the goal of this PR?

Document adding concepts to an existing schema. 
Closes https://github.com/vaticle/docs/issues/650.

## What are the changes implemented in this PR?

Created a modify section with add and delete concepts to/from a schema.
Added a section about adding concepts to a schema. 
Moved an undefine section as delete into the modify.

izmalk committed:
Added Add to a schema

Created a modify section with add and delete.
Added a section about adding concepts to a schema.
Moved an undefine section as delete into the modify.

izmalk committed:
Fix for query checks

Changed attribute from address to phone-number to pass the query checks

izmalk committed:
Fix undefine links

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+## Modify an existing schema
+We can modify an existing schema by adding or deleting concepts: entities, relations, attributes and rules.
+
+To add we use `define` and to delete an existing concept we use `undefine`. 
```
```suggestion
To add a concept we use `define` and to delete an existing concept we use `undefine`. 
```

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+shall delete the old concept (by using [undefine](#deleting-from-a-schema-with-undefine)) and add the new one (by using 
+[define](#adding-to-a-schema-with-define)).
+
+Don't forget to modify all references accordingly (with deletion and adding).
```
```suggestion
Don't forget to update all references to the modified concept through the same process (i.e. deletion and addition).
```

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+
+Don't forget to modify all references accordingly (with deletion and adding).
+
+If there are a lot of changes to be done consider creating a brand-new database to load a new schema from scratch.
```
It might be the case that creating a new database would involve also migrating all of the data (could be very painful) - perhaps we could advise that the user make use of Studio as it supports schema editor actions without starting from scratch.

izmalk replied:
Never heard of such an ability/function in Studio.
Most pains in modifying schema are from the need to delete/modify sub-types and data. 
So deleting all data might be a by-product of creating a new DB, but it is often a necessity.
Let's add a note here about adding data from scratch and leave it that way. 
And I want to know more about this Studio schema editor.

izmalk replied:
I think I managed to do both with minimal changes. https://github.com/vaticle/docs/pull/652#:~:text=85020f7-,Fix%20after%20review,-%E2%80%A6

jamesreprise commented on 09-schema/01-concepts.md:
````diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+```
+</div>
+
+In the example above the `person` was already defined as a subtype of an entity as well as first three attributes.
````
```suggestion
In the example above, `person` was already defined as a subtype of `entity` and owning the first three attributes.
```

jamesreprise commented on 09-schema/01-concepts.md:
````diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+```
+</div>
+
+Any concepts that already exist in a schema you are trying to modify will not be added to prevent doubling. That is why 
````
doubling -> duplication

This applies everywhere else we've used the word 'doubling'. Not necessarily wrong, just more familiar language to most developers imo.

izmalk replied:
I think you are right here. Fixed it everywhere I found it  https://github.com/vaticle/docs/pull/652/commits/85020f7fcfe6cf2e6378f3e1e7c6d38cfd9b10b7

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+You can even reuse your original query/file (e.g. `schema.tql`) that was used to create the original schema. 
+By adding concepts to the original query and executing it again you will achieve the same result — any concepts that 
+were created by the original request will not be created again (doubled) but the newly added concepts will be added to 
+the schema. Thus, running the same query a second time will have no effect.
```
```suggestion
the schema. Thus, running the same query a second time will have no effect (i.e. it is idempotent).
```

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+
+<div class="note">
+[Important]
+You can only add concepts this way. Do not try to modify existing concepts or delete by modifying query and launching 
```
```suggestion
You can only add concepts this way. Do not try to modify existing concepts or delete by modifying the query and launching 
```

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -751,16 +751,90 @@ TypeQLDefine query = TypeQL.define(
+### Adding to a schema with Define
+
+To add concepts to an existing schema we need to do a query with a single `define` statement at the beginning, just like
+for a new schema. For example, lets add attribute `phone-number` to the `person` entity:
```
lets -> let's

izmalk replied:
```suggestion
for a new schema. For example, let's add attribute `phone-number` to the `person` entity:
```

izmalk committed:
Apply suggestions from code review

Co-authored-by: James Williams <james@berserksystems.com>

izmalk committed:
Fix typo in 09-schema/01-concepts.md

izmalk committed:
Renamed doubling to duplication

izmalk committed:
Fix after review

Added a mention of the Studio and a note on load data into the new database.

izmalk committed:
Fix a link

Initially, I was going to leave it empty and add a correct link later with the new structure. But I forgot this is a separate PR and branch. So I fixed the link as in the old doc structure.

izmalk committed:
Minor fixes

Fixed using you instead of we.
Refactored notes. Double-checked Idempotence.

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -752,22 +752,38 @@ TypeQLDefine query = TypeQL.define(
-We can modify an existing schema by adding or deleting concepts: entities, relations, attributes and rules.
 
-To add we use `define` and to delete an existing concept we use `undefine`. 
+We can modify an existing database schema. Usually it is done by adding or deleting types: entities, relations, 
```
```suggestion
We can modify an existing database schema. Usually this is done by adding or deleting types: entities, relations, 
```

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -812,25 +828,27 @@ TypeQLDefine query = TypeQL.define(
-You can only add concepts this way. Do not try to modify existing concepts or delete by modifying query and launching 
-it again.
+Since this operation is idempotent we can add any amount of previously applied statements that are already in the schema 
+into a define request. But what we can only add types this way. We can't modify or delete existing types by modifying 
```
```suggestion
into a define request. But, we can only add types this way. We can't modify or delete existing types by modifying 
```

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -812,25 +828,27 @@ TypeQLDefine query = TypeQL.define(
 <div class="note">
 [Important]
-Don't forget to `commit` after executing an `undefine` statement. Otherwise, anything you have undefined is NOT committed to the original database that is running on the TypeDB server.
+Don't forget to `commit` after executing an `undefine` statement. Otherwise, anything we have undefined is NOT committed to the original database that is running on the TypeDB server.
```
```suggestion
Don't forget to `commit` after executing an `undefine` query. Otherwise, any changes will NOT be committed to the database.
```

izmalk replied:
If I understand correctly, there can be (and usually are) multiple queries in a define/undefine tql file. All of them go with a single transaction, but different queries. So what would we call this "text of the tql file with define in it"?

izmalk replied:
Also, are we committing the state of the query? I mean is that a correct term? I thought we more like commit changes made by the query.

jamesreprise replied:
> If I understand correctly, there can be (and usually are) multiple queries in a define/undefine tql file. All of them go with a single transaction, but different queries. So what would we call this "text of the tql file with define in it"?

I believe the whole file, if it is a single define statement, is one query.



> Also, are we committing the state of the query? I mean is that a correct term? I thought we more like commit changes made by the query.

You're right, this is poorly worded. Perhaps 'any changes will NOT [...]'. No need to go into more detail.

jamesreprise commented on 09-schema/01-concepts.md:
```diff
@@ -16,7 +16,7 @@ To try the following examples with one of the TypeDB clients, follows these [Cli
 <div class="note">
 [Important]
-Don't forget to `commit` after executing a `define` query. Otherwise, anything you have defined is NOT committed to the original database that is running on the TypeDB server.
+Don't forget to `commit` after executing a `define` query. Otherwise, anything we have defined is NOT committed to the original database that is running on the TypeDB server.
```
```suggestion
Don't forget to `commit` after executing a `define` query. Otherwise, any changes will NOT be committed to the database.
```

izmalk committed:
Apply suggestions from code review

Co-authored-by: James Williams <james@berserksystems.com>

izmalk committed:
Apply suggestions from code review

Co-authored-by: James Williams <james@berserksystems.com>

alexjpwalker reviewed (changes requested):
Reviewed in person. Section on mutating existing concepts needs minor work.

izmalk committed:
Update after review with Alex
